### PR TITLE
[SPARK-34614][SQL] ANSI mode: Casting String to Boolean should throw exception on parse error

### PR DESCRIPTION
**Pull request by gengliangwang**
### What changes were proposed in this pull request?

In ANSI mode, casting String to Boolean should throw an exception on parse error, instead of returning null

### Why are the changes needed?

For better ANSI compliance 

### Does this PR introduce _any_ user-facing change?

Yes, in ANSI mode there will be an exception on parse failure of casting String value to Boolean type.

### How was this patch tested?

Unit tests.

### CODE REVIEW

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala [hunk 1]
```diff
@@ -343,168 +343,6 @@ abstract class CastSuiteBase extends SparkFunSuite with ExpressionEvalHelper {
     checkCast(Decimal(1.5), "1.5")
   }
 
-  test("cast from array") {
-    val array = Literal.create(Seq("123", "true", "f", null),
-      ArrayType(StringType, containsNull = true))
-    val array_notNull = Literal.create(Seq("123", "true", "f"),
-      ArrayType(StringType, containsNull = false))
-
-    checkNullCast(ArrayType(StringType), ArrayType(IntegerType))
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala:352]: can we leave the positive tests in the base suite?

**Comment by cloud-fan** [comment]
let's also update `docs/sql-ref-ansi-compliance.md`

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala (hunk 1 above)
Comment (gengliangwang) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala:352]: Thanks, updated.

File: docs/sql-ref-ansi-compliance.md [hunk 2]
```diff
@@ -165,6 +165,7 @@ The behavior of some SQL operators can be different under ANSI mode (`spark.sql.
   - `map_col[key]`: This operator throws `NoSuchElementException` if key does not exist in map.
   - `CAST(string_col AS TIMESTAMP)`: This operator should fail with an exception if the input string can't be parsed.
   - `CAST(string_col AS DATE)`: This operator should fail with an exception if the input string can't be parsed.
+  - `CAST(string_col AS BOOLEAN)`: This operator should fail with an exception if the input string can't be parsed.
```
Comment (cloud-fan) [inline docs/sql-ref-ansi-compliance.md:168]: not related to this PR. Does section `### Type Conversion` cover the CAST behaviors?

Comment (gengliangwang) [reply docs/sql-ref-ansi-compliance.md:168]: Yes and no.
```
When `spark.sql.ansi.enabled` is set to `true`, explicit casting by `CAST` syntax throws a runtime exception for illegal cast patterns defined in the standard, e.g. casts from a string to an integer.
```
covers this, but we had better write down all the cases that would throw exceptions.

How about a follow-up to add a subsection for all these cases under `### Type Conversion`?

Comment (cloud-fan) [reply docs/sql-ref-ansi-compliance.md:168]: SGTM

**Review by cloud-fan — APPROVED** [review]

**Comment by gengliangwang** [comment]
Merging to master

**Comment by maropu** [comment]
late lgtm

**Pull request closed without merging**